### ISSUE TITLE
Memory consumption

### ISSUE THREAD

**Issue by iXanthos** [openml-labs/gama#92]
Hello, 

I would like to ask about the memory consuption of `gama`. At `autosklearn`, there are parameters ( `ml_memory_limit` and `ensemble_memory_limit` ) that allow the user to setup the amount of memory the automl uses. I searched the docs of `gama` but could not find such options and the only mention of `memory` can be found at the Release notes (`Memory usage of all GAMA’s processes is logged.`). Where is it logged exactly? I guess at `gama.log`, but searching for `memory` returned no results. What is the general policy of memory consumption of gama? I am asking because I will be performing experiments on an HPC and I would like to know how much memory I need to allocate to the task, in order to have `gama` use ~50GBs.

Thank you.

**Comment by PGijsbers** [comment]
Hi! This is something I have only recently started working on, but had to put off because other things came up. It's currently in early stages, but it's the first thing I will address when I have time. I expect it to be finished and tested in 1~2 weeks. If you can wait that long, I would recommend to do so.

> Where is it logged exactly?

They are indeed in `gama.log`, you will find lines starting with an `M` token.
E.g.:

> M,2020-05-07T22:40:22.510430,6552,112.2109375,13760,103.69921875,2172,103.91796875,2228,104.1796875,8024,103.89453125

where the format is: 
 - 'M', token 
 - Datetime in ISO format
 - (pid, memory)-pairs that denote the memory usage in mb for each of GAMA's processes. The first pair is the main process, the other pairs are the subprocesses that are evaluating pipelines.
(side note: I don't know how long text logs will be around for)

> What is the general policy of memory consumption of gama?

By default, none. In practice this works because excessive memory usage only occurs during a pipeline evaluation where MemoryErrors are "expected" (both by being in a subprocess, and raised from a try/except-clause). The goal for the new behavior that allows you to impose a limit, is for the user to be able to specify a memory limit for GAMA, and GAMA takes care of ensuring the constraint is met including its subprocesses.

**Comment by iXanthos** [comment]
> Hi! This is something I have only recently started working on, but had to put off because other things came up. It's currently in early stages, but it's the first thing I will address when I have time. I expect it to be finished and tested in 1~2 weeks. If you can wait that long, I would recommend to do so.
I see, I would be interested in using that, I may start experimenting with the current version and rerun the experiments with this added functionality.

**Comment by PGijsbers** [comment]
I'll post here when I've made the changes!

**Comment by PGijsbers** [comment]
I think I'm very close. It seems to behave nicely except for an infrequent error. I am investigating it and I am hopeful that I will have it merged with `develop` Monday latest (unfortunately I can't work tomorrow).

**Comment by iXanthos** [comment]
Thank you for looking into it. I have run some experiments on both binary classification and regression tasks and gama reaches a 97% and 98% completion rate. To compare, autosklearn's completion rates are lower (96% and 93%), so I think at least in terms of stability it looks great.

**Comment by PGijsbers** [comment]
That's great to hear! I really want to keep it that way :)
Unfortunately, the cluster I use for tests was rather busy today so I did not get as many iterations done as I'd hoped. I'm still hopeful it will be done tomorrow.

**Comment by PGijsbers** [comment]
It has been merged into develop. But I am not completely happy with it. I will slowly working on it while I do other stuff. In the meantime, it should be mostly functional :) use the `max_memory_mb` parameter when initializing gama.

**Comment by iXanthos** [comment]
I see, nice work!
A question on this matter, at this point (without setting the `max_memory_mb` arg), is gama defaulting to a specific amount of RAM? In the logs I see that for a model it uses 
`125669,104.84375,125711,106.69921875,125713,111.53125,125717,109.88671875,125721,133.44921875,125726,125.55078125,125731,126.0234375,125736,107.703125,125741,122.8984375`, which is approximately ~1GB of memory. I guess if it doesnt need to, it will not use all the memory available.

Also, does gama reserve the memory required  for each model and once the model is finished, it flushes it (or at least a part of it) in order for it to be available for the next models? Or does it split the total amount of memory, depending on the number of models created?

**Comment by PGijsbers** [comment]
By default, no memory limit is set and GAMA will be able to use up to the maximum memory your python process may have (based on e.g. OS settings) for each of the `n_jobs`. Memory usage will then depend on factors such as the dataset or pipeline being evaluated.

I am not entirely sure what you are asking with the second question. But I can give you a rundown which might answer it anyway. First we consider the scenario with no memory constraint.
For search, GAMA starts `n_jobs` subprocesses that evaluate pipelines on the dataset. The pipelines to evaluate are generated in the main process (i.e. which steps to use and their configuration) and sent to a joint queue from which the evaluation subprocesses can unqueue. Each subprocess will then unqueue a pipeline and evaluate it on the dataset. It returns the trained pipelines and their predictions to the main process. The main process will record the evaluation. If the pipeline is good enough to be considered for the final model, the pipeline (and possibly its predictions) is saved to disk. The predictions and trained pipeline are then removed from memory.

With a set memory constraint, the above process is pretty much the same. Additionally, GAMA will inspect the memory usage of each subprocess (and the main process) at regular intervals (currently 50ms but this can probably be relaxed). If this total violates the memory constraint, the subprocess with the highest memory usage is killed. If violations happen infrequently (currently less than 1 in 20 evaluations), a new subprocess is started in its place. If violations happen too frequently, no new subprocess is started. Over time the goal of this policy is to effectively downscales `n_jobs` to an amount that can make the best use of the memory without violating the constraint.

It is very coarse but simple, which I think is good for now. It's far from perfect and there are still issues I need to work out. But in the long term instead I hope to be able to use a mature package for evaluating pipelines asynchronously and keeping with memory constraints. Unfortunately other things have to take priority right now since it probably requires a lot of work while the current implementation is functional enough.

Let me know if that answered your question or if you have more.

**Comment by PGijsbers** [comment]
@iXanthos did you ever try out the feature to constrain memory usage?

**Comment by iXanthos** [comment]
> @iXanthos did you ever try out the feature to constrain memory usage?

Hey, sorry for the really late reply, I am on mandatory army duty, hence I couldnt connect with you earlier. Yes, this solution fixed some of my issues, but as I recall, the tool crashed when it had to deal with data with many features (over 5000). Unfortunately I cannot test it myself atm. 

Regards,
Iordanis

**Comment by PGijsbers** [comment]
Thanks for letting me know! I don't know if I'll have the time to look into that short term, but I'll keep the issue open as a reminder to look into memory management/stability for datasets with a high number of features.

**Comment by PGijsbers** [comment]
Closing this issue. If you experience issues with specific large data, please open a new issue with a minimal example :)
